### ISSUE TITLE
SpaceAroundOperator: false positives for `[…|…]`

### ISSUE THREAD

**Issue by NobbZ** [rrrene/credo#477]
### Environment

* Credo version (`mix credo -v`): 0.8.10
* Erlang/Elixir version (`elixir -v`):
```
Erlang/OTP 20 [erts-9.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe]

Elixir 1.5.2
```
* Operating system: Funtoo Linux

### What were you trying to do?

Linting a single file ;)

Also I was unable to extract a simpler example, therefore I have to post my solution to Advent of Code 17 here, I do behind spoiler tags though:

<details><summary>Code with false positives</summary>

```ex
defmodule AoC.Day9 do
  @moduledoc false

  use AoC.Default, into: :default

  @default @default
  |> String.trim()
  |> String.codepoints()

  def a(stream \\ @default), do: score_groups(stream)

  def b(stream \\ @default), do: count_garbage(stream)

  defp score_groups(stream, state \\ [], curr \\ 0, sum \\ 0)
  defp score_groups([], [], _, sum), do: sum
  defp score_groups(["{"|t], [], 0, sum), do: score_groups(t, [:group], 1, sum + 1)
  defp score_groups(["<"|t], [], 0, sum), do: score_groups(t, [:garbage], 0, sum)
  defp score_groups(["{"|t], [:group|s], curr, sum), do: score_groups(t, [:group, :group|s], curr + 1, sum + curr + 1)
  defp score_groups(["}"|t], [:group|s], curr, sum), do: score_groups(t, s, curr - 1, sum)
  defp score_groups(["<"|t], [:group|s], curr, sum), do: score_groups(t, [:garbage, :group|s], curr, sum)
  defp score_groups([","|t], [:group|s], curr, sum), do: score_groups(t, [:group|s], curr, sum)
  defp score_groups(["!", _|t], [:garbage|s], curr, sum), do: score_groups(t, [:garbage|s], curr, sum)
  defp score_groups([">"|t], [:garbage|s], curr, sum), do: score_groups(t, s, curr, sum)
  defp score_groups([_|t], [:garbage|s], curr, sum), do: score_groups(t, [:garbage|s], curr, sum)

  defp count_garbage(stream, garbage \\ false, chars \\ 0)
  defp count_garbage([], false, chars), do: chars
  defp count_garbage(["<"|t], false, chars), do: count_garbage(t, true, chars)
  defp count_garbage([_|t], false, chars), do: count_garbage(t, false, chars)
  defp count_garbage(["!", _|t], true, chars), do: count_garbage(t, true, chars)
  defp count_garbage([">"|t], true, chars), do: count_garbage(t, false, chars)
  defp count_garbage([_|t], true, chars), do: count_garbage(t, true, chars + 1)
end
```

</details>

### Expected outcome

No warnings

### Actual outcome

The code above produces the following warnings (but only if linted as a single file, not when linting the project):

```plain
Checking 1 source file ...

  Consistency
┃
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:32:76 #(AoC.Day9.count_garbage)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:19:82 #(AoC.Day9.score_groups)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:18:99 #(AoC.Day9.score_groups)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:18:108 #(AoC.Day9.score_groups)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:18:115 #(AoC.Day9.score_groups)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:16:80 #(AoC.Day9.score_groups)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:12:16 #(AoC.Day9.b)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:10:16 #(AoC.Day9.a)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:8:3 #(AoC.Day9)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:7:3 #(AoC.Day9)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:26:38 #(AoC.Day9.)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:26:54 #(AoC.Day9.)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:14:35 #(AoC.Day9.)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:14:47 #(AoC.Day9.)
┃ [C] ↗ There are no spaces around operators most of the time, but here there
┃       are.
┃       lib/aoc/day9.ex:14:57 #(AoC.Day9.)

Please report incorrect results: https://github.com/rrrene/credo/issues

Analysis took 0.09 seconds (0.00s to load, 0.09s running checks)
21 mods/funs, found 15 consistency issues.
```

By applying the following patch I can make the warnings go away, but have a much less consistent code (spoiler agains since it contains parts of my solution).

<details><summary>The patch removing the warnings</summary>

```diff
diff --git a/lib/aoc/day9.ex b/lib/aoc/day9.ex
index e21cf72..3f620b4 100644
--- a/lib/aoc/day9.ex
+++ b/lib/aoc/day9.ex
@@ -13,12 +13,12 @@ defmodule AoC.Day9 do

   defp score_groups(stream, state \\ [], curr \\ 0, sum \\ 0)
   defp score_groups([], [], _, sum), do: sum
-  defp score_groups(["{"|t], [], 0, sum), do: score_groups(t, [:group], 1, sum + 1)
-  defp score_groups(["<"|t], [], 0, sum), do: score_groups(t, [:garbage], 0, sum)
-  defp score_groups(["{"|t], [:group|s], curr, sum), do: score_groups(t, [:group, :group|s], curr + 1, sum + curr + 1)
-  defp score_groups(["}"|t], [:group|s], curr, sum), do: score_groups(t, s, curr - 1, sum)
-  defp score_groups(["<"|t], [:group|s], curr, sum), do: score_groups(t, [:garbage, :group|s], curr, sum)
-  defp score_groups([","|t], [:group|s], curr, sum), do: score_groups(t, [:group|s], curr, sum)
+  defp score_groups(["{" | t], [], 0, sum), do: score_groups(t, [:group], 1, sum + 1)
+  defp score_groups(["<" | t], [], 0, sum), do: score_groups(t, [:garbage], 0, sum)
+  defp score_groups(["{" | t], [:group|s], curr, sum), do: score_groups(t, [:group, :group|s], curr + 1, sum + curr + 1)
+  defp score_groups(["}" | t], [:group|s], curr, sum), do: score_groups(t, s, curr - 1, sum)
+  defp score_groups(["<" | t], [:group|s], curr, sum), do: score_groups(t, [:garbage, :group|s], curr, sum)
+  defp score_groups(["," | t], [:group|s], curr, sum), do: score_groups(t, [:group|s], curr, sum)
   defp score_groups(["!", _|t], [:garbage|s], curr, sum), do: score_groups(t, [:garbage|s], curr, sum)
   defp score_groups([">"|t], [:garbage|s], curr, sum), do: score_groups(t, s, curr, sum)
   defp score_groups([_|t], [:garbage|s], curr, sum), do: score_groups(t, [:garbage|s], curr, sum)
```

</details>

**Comment by rrrene** [comment]
Hi @NobbZ and thanks for the report!

I am unsure why this is considered wrong behaviour. `|` is considered a normal operator here, just like `+` and `*`. Are you proposing that it is always `[h|t]` and never `[h | t]`?

**Comment by NobbZ** [comment]
For me `|` is strictly different to any other operator, since I do not see `|`, but `[…|…]` as the operator.

Even worse, when introducing inconsistencies, then the warning goes away.

Look at the diff attached.

When I change function heads from `foo([a|b], [c|d])`-style to `foo([a | b], [c|d])` the warning is **gone**.

So, aside from the fact that I prefer no spaces around the vertical bar, it seems as if the scanner is not consistent.

I can live with space around the bar, as far as I remember, `mix format` will add it anyway, but we should probably keep this open because of the inconsistency and investigate further?

**Comment by rrrene** [comment]
> we should probably keep this open because of the inconsistency and investigate further?

You are right :+1:

**Comment by bozydar** [comment]
@rrrene 
I removed the `|` from the `@default_params[:ignored]` list in the `Credo.Check.Consistency.SpaceAroundOperators` and it started working consistent. The problem is how to handle this code:
```
  @without_spaces """
  defmodule Credo.Sample1 do
    defmodule InlineModule do
      @type config_or_func :: Config.t() | (-> Config.t())
    end
  end
  """
```
The type definition's `|` should always be with-spaces imho. Do you think it is better to have `|` always with spaces regardles the context OR rather to check if in the `::` was before in the same line?
Imho the first option is better because `Code.format_string!/2` inserts the spaces as @NobbZ has mentioned.

**Comment by NobbZ** [comment]
As some formatting rules are enforced by the official formatter, I think we should transition credo to be fully consistent with them, even if this means to make some consistency checks to hard rules…

**Comment by rrrene** [comment]
> we should transition credo to be fully consistent with [the official formatter]

:+1:

@josevalim opened PR #443 a while back, showcasing which checks could be removed in the future.

I think we should try to follow @NobbZ's idea and - instead of removing them - make sure that we can make these consistent with Elixir 1.6 formatting behaviour:

```
lib/credo/check/consistency/line_endings.ex
lib/credo/check/consistency/line_endings/collector.ex
lib/credo/check/consistency/space_around_operators.ex
lib/credo/check/consistency/space_around_operators/collector.ex
lib/credo/check/consistency/space_around_operators/space_helper.ex
lib/credo/check/consistency/space_in_parentheses.ex
lib/credo/check/consistency/space_in_parentheses/collector.ex
lib/credo/check/consistency/tabs_or_spaces.ex
lib/credo/check/consistency/tabs_or_spaces/collector.ex
lib/credo/check/readability/large_numbers.ex
lib/credo/check/readability/max_line_length.ex
lib/credo/check/readability/parentheses_in_condition.ex
lib/credo/check/readability/redundant_blank_lines.ex
lib/credo/check/readability/semicolons.ex
lib/credo/check/readability/space_after_commas.ex
lib/credo/check/readability/trailing_blank_line.ex
lib/credo/check/readability/trailing_white_space.ex
```

@NobbZ @bozydar Is this a good idea or is removing the checks the cleaner approach after all?WDYT?

**Comment by josevalim** [comment]
One option is to move those to a new place, called code_style or something, and disable them by default for developers running on v1.6. Another option is to move them to a separate project and people can depend on it if they want those checks.

**Comment by rrrene** [comment]
@josevalim We could easily "remove" them by disabling them for Elixir versions `>= 1.6.0` (that is already an option in Credo). But is it correct to assume that everyone running Elixir 1.6 IS actually using the formatter?

**Comment by josevalim** [comment]
@rrrene you are right, I don't think it is safe to assume that :)

However, if you want any consistency on those elements, the formatter is the best way to get it: it is automatic and it has no false positives. The formatter also says about things that it is probably not worth trying to make consistent, such as the "starts with a pipeline" check that I proposed myself, and today I consider a mistake. :) Credo beautifully complements the formatter by helping developers focus on the remaining issues, which go beyond the syntax and are more important to developers.

I assume Credo is not going to be able to run away from 1.0 for much longer. :) It started when Elixir was on version 1.1 and it has become a great asset for the community. Meanwhile, the community and ecosystem also changed. And now with the formatter, it may be worth looking inside and have the Credo team answer where you think Credo will give the most value to developers and where you want to put your efforts towards 1.0 on.

With Elixir v1.6, my opinion is that Credo is the most useful when it goes beyond code style. There are some consistency checks that are still important, like the variable names, but I would prefer to see less focus on that area. It feels like the next step for the community is to focus more on code quality measurements and Credo can once more lead by example. If you move the existing formatting checks to a separate project, developers who dislike the code formatter will still have an option to run them if they want to, and they can help maintain the "old checks" while the Credo team focuses on the new ones.

If it is of any interest, I can review the checks that are "deprecated" with the formatter to help move the discussion forward.

**Comment by rrrene** [comment]
@josevalim This is funny. I am just now writing a "roadmap" for Credo for 2018, outlining something similar :)

And I had a complementing thought since christmas: 

> What if we decide that Credo `v1.0` needs Elixir `>= 1.6`? 

Not out of technical necessity, but to make a clean cut and move forward with a cleaner separation between Elixir's own tooling (compiler warnings, formatting, etc.) and Credo's linting annotations.

I would not have put it in the same words as you did, but I think we are agreeing on the future direction for Credo.

The areas I want to focus on in 2018 are:

* improve APIs for code analysis (i.e. make Credo's internal tooling and documentation better)
* improve Editor integration
* improve Speed (at some point we might want to introduce a Credo Server or some other caching mechanism)
* Plugin architecture (this is probably for v1.1 or later)
* Ecto, Phoenix, and Nerves (there could be plugins for popular frameworks and libs)

In this scenario, Credo `v0.9.x` would be the last version to support Elixir `< 1.6` and the last version to include the checks we are discussing here (and I think that we all would love your input regarding which checks should make the cut for `v1.0`).

WDYT?

P.S.

> Credo beautifully complements the formatter by helping developers focus on the remaining issues, which go beyond the syntax and are more important to developers.

I want to print that on a t-shirt 💜

**Comment by josevalim** [comment]
I think requiring v1.6, which is barely a month old, is probably too abrupt since you don't plan to gain anything from it. I think removing the style-based checks is enough to send the message you want while still allowing people to use older versions.

> Not out of technical necessity, but to make a clean cut and move forward with a cleaner separation between Elixir's own tooling (compiler warnings, formatting, etc.) and Credo's linting annotations.

I think this is also a very good point. We have made the compiler warnings smarter so being able to sync on those to avoid redundancy can also be useful.

**Comment by josevalim** [comment]
@rrrene https://github.com/rrrene/credo/issues/505 :)

**Comment by rrrene** [comment]
@NobbZ Hi, thx again for this. Unfortunately, I am closing it due to "it being no longer applicable" in the latest version of Credo (IMHO, that is). 

If you disagree, please re-open and elaborate! :+1:
